File: core/examples/social/track_whale.py
import signal
import sys
from datetime import datetime

import pmxt


def fmt(address: str) -> str:
    if len(address) > 10:
        return f"{address[:6]}...{address[-4:]}"
    return address


def run():
    exchange = pmxt.Polymarket()

    # -- Step 1: Define and find whales ----------------------------------------
    # For simplicity, here I assume the trader with the largest volume is a "whale"
    print("Fetching top volume traders in all time...\n")
    whales = exchange.call_api("getV1Leaderboard", {
        "category": "OVERALL",
        "timePeriod": "ALL",
        "orderBy": "VOL",
        "limit": 10,
    })

    print("Rank  Name                           Address        Volume (USDC)   PnL (USDC)")
    print("-" * 82)
    for w in whales:
        rank = str(w.get("rank", "")).rjust(2)
        name = fmt(w.get("userName", "")).ljust(30)
        addr = fmt(w.get("proxyWallet", "")).ljust(14)
        vol  = f"${w.get('vol', 0) / 1_000_000:.1f}M".rjust(14)
        pnl  = f"${w.get('pnl', 0) / 1_000:.1f}K".rjust(12)
        print(f"  {rank}  {name} {addr} {vol} {pnl}")

    # -- Step 2: Watch the top whale -------------------------------------------
    whale = whales[0]
    label = whale.get("userName")
    address = whale.get("proxyWallet")
    print(f"\nWatching {label} ({address}) ...")
    print("Press Ctrl+C to stop.\n")

    running = True

    def handle_sigint(sig, frame):
        nonlocal running
        print("\nStopping...")
        running = False
        try:
            exchange.unwatch_address(address)
        except Exception:
            pass
        finally:
            sys.exit(0)

    signal.signal(signal.SIGINT, handle_sigint)

    try:
        while running:
            snapshot = exchange.watch_address(address, ["trades", "positions"])
            print(f"\n[Update @ {datetime.now().strftime('%H:%M:%S')}]")
            if snapshot.trades:
                for t in snapshot.trades:
                    print(f"  Trade: {t.side.upper()} {t.amount:.0f} shares @ ${t.price:.3f}")
            if snapshot.positions:
                for p in snapshot.positions:
                    print(f"  Position: {p.outcome_label} size={p.size:.2f} entry=${p.entry_price:.3f}")
    except Exception as e:
        print(f"Error: {e}")


if __name__ == "__main__":
    run()

File: core/examples/social/track_whale.ts
import pmxt from '../../src';

function fmt(address: string): string {
    if (address.length > 10) {
        return `${address.slice(0, 6)}...${address.slice(-4)}`;
    }
    return address;
}

async function run() {
    const exchange = new pmxt.Polymarket();

    // -- Step 1: Define and find whales -----------------------------------------------
    // For simplicity, here I assume the trader with the largest volume is a "whale"
    console.log('Fetching top volume traders in all time...\n');
    const whales: any[] = await (exchange as any).getV1Leaderboard({
        category: 'OVERALL',
        timePeriod: 'ALL',
        orderBy: 'VOL',
        limit: 10,
    });

    console.log('Rank  Name                           Address        Volume (USDC)   PnL (USDC)');
    console.log('-'.repeat(82));
    for (const w of whales) {
        const name = fmt(w.userName ?? '').padEnd(30);
        const addr = fmt(w.proxyWallet ?? '').padEnd(14);
        const vol = `$${((w.vol ?? 0) / 1_000_000).toFixed(1)}M`.padStart(14);
        const pnl = `$${((w.pnl ?? 0) / 1_000).toFixed(1)}K`.padStart(12);
        console.log(`  ${String(w.rank ?? '').padStart(2)}  ${name} ${addr} ${vol} ${pnl}`);
    }

    // -- Step 2: Watch the top whale --------------------------------------------------
    const whale = whales[0];
    const label = whale.userName;
    const address = whale.proxyWallet;
    console.log(`\nWatching ${label} (${address}) ...`);
    console.log('Press Ctrl+C to stop.\n');

    let running = true;
    process.on('SIGINT', async () => {
        console.log('Stopping...');
        running = false;
        await exchange.unwatchAddress(address);
        process.exit(0);
    });

    try {
        while (running) {
            const snapshot = await exchange.watchAddress(address, ['trades', 'positions']);
            console.log(`\n[Update @ ${new Date().toLocaleTimeString()}]`);
            if (snapshot.trades) {
                for (const t of snapshot.trades) {
                    console.log(`  Trade: ${t.side.toUpperCase()} ${t.amount.toFixed(0)} shares @ $${t.price.toFixed(3)}`);
                }
            }
            if (snapshot.positions) {
                for (const p of snapshot.positions) {
                    console.log(`  Position: ${p.outcomeLabel} size=${p.size.toFixed(2)} entry=$${p.entryPrice.toFixed(3)}`);
                }
            }
        }
    } catch (err) {
        console.error(err);
    }
}

run().catch(console.error);
